lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,783 @@
+/// may be denominated in currency other than bitcoin but are ultimately paid using the latter.
+///
+/// Through the use of [`BlindedPath`]s, offers provide recipient privacy.
+#[derive(Clone, Debug)]
+pub struct Offer {
+	// The serialized offer. Needed when creating an `InvoiceRequest` if the offer contains unknown
+	// fields.
+	bytes: Vec<u8>,
+	contents: OfferContents,
+}
+
+/// The contents of an [`Offer`], which may be shared with an `InvoiceRequest` or an `Invoice`.
+#[derive(Clone, Debug)]
+pub(crate) struct OfferContents {
+	chains: Option<Vec<ChainHash>>,
+	metadata: Option<Vec<u8>>,
+	amount: Option<Amount>,
+	description: String,
+	features: OfferFeatures,
+	absolute_expiry: Option<Duration>,
+	issuer: Option<String>,
+	paths: Option<Vec<BlindedPath>>,
+	quantity_min: Option<u64>,
+	quantity_max: Option<u64>,
+	signing_pubkey: Option<PublicKey>,
+}
+
+impl Offer {
+	// TODO: Return a slice once ChainHash has constants.
+	/// The chain used for paying the invoice.
+	pub fn chains(&self) -> Vec<ChainHash> {
+		self.contents.chains
+			.as_ref()
+			.cloned()
+			.unwrap_or_else(|| vec![ChainHash::using_genesis_block(Network::Bitcoin)])
+	}
+
+	/// Metadata set by the originator. Useful for authentication and validating fields.
+	pub fn metadata(&self) -> Option<&Vec<u8>> {
+		self.contents.metadata.as_ref()
+	}
+
+	/// The minimum amount required for a successful payment.
+	pub fn amount(&self) -> Option<&Amount> {
+		self.contents.amount.as_ref()
+	}
+
+	/// A complete description of the purpose of the payment.
+	pub fn description(&self) -> &str {
+		&self.contents.description
+	}
+
+	/// Features for paying the invoice.
+	pub fn features(&self) -> &OfferFeatures {
+		&self.contents.features
+	}
+
+	/// Duration since the Unix epoch when an invoice should no longer be requested.
+	///
+	/// If `None`, the offer does not expire.
+	pub fn absolute_expiry(&self) -> Option<Duration> {
+		self.contents.absolute_expiry
+	}
+
+	/// Whether the offer has expired.
+	#[cfg(feature = "std")]
+	pub fn is_expired(&self) -> bool {
+		match self.absolute_expiry() {
+			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
+				Ok(elapsed) => elapsed > seconds_from_epoch,
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	/// The issuer of the offer, possibly beginning with `user@domain` or `domain`.
+	pub fn issuer(&self) -> Option<&str> {
+		self.contents.issuer.as_ref().map(|issuer| issuer.as_str())
+	}
+
+	/// Paths to the recipient originating from publicly reachable nodes.
+	pub fn paths(&self) -> &[BlindedPath] {
+		self.contents.paths.as_ref().map(|paths| paths.as_slice()).unwrap_or(&[])
+	}
+
+	/// The minimum quantity of items supported.
+	pub fn quantity_min(&self) -> u64 {
+		self.contents.quantity_min()
+	}
+
+	/// The maximum quantity of items supported.
+	pub fn quantity_max(&self) -> u64 {
+		self.contents.quantity_max()
+	}
+
+	/// The recipient's public key used to sign invoices.
```

Comment:
Per Matt's earlier comment, may be better to phrase it as "The public key used by the recipient to sign invoices"